### PR TITLE
Pp 607/place transactions around cancel

### PR DESCRIPTION
**Pull request by ckalista**
- Common interface GatewayRequest
  - New common interface for AuthorisationRequest, CancelRequest and CaptureRequest.
  - getPaymentProviderFor method in CardService abstract class to retrieve the provider for a given charge entity.
- CardService abstract class generalise preOperation
  - Introduce new charge statuses CANCEL_READY and CANCEL_ERROR.
  - Tighten up CancelResponse to force the object to be created with the ErrorResponse and ChargeStatus set.
  - Reinstate CardService abstract class which introduces a hierarchy between CardAuthoriseService, CardCancelService and CardCaptureService with the purpose to share common preOperation logic.
  - Move hasStatus method to ChargeEntity where it belongs.
  - Map CANCEL_READY as IN_PROGRESS external status.
  - Map CANCEL_ERROR as FAILED external status.
- Refactor CardService abstract class into TransactionalGatewayOperation Interface
  - Refactor CardService abstract class into an Interface called TransactionalGatewayOperation with has one default method implementation executeGatewayOperation taking care of orchestrating the operations: preOperation, operation and postOperation.
  - Refactor CardAuthorisationService, CardCaptureService and CardCancelService to comply to the new TransactionalGatewayOperation interface.

### CODE REVIEW

**Comment by imranbohoran** [comment]
rebuild

File: src/main/java/uk/gov/pay/connector/dao/ChargeDao.java [hunk 1]
```diff
@@ -88,7 +88,7 @@ public void persist(ChargeEntity chargeEntity) {
         return query.getResultList();
     }
 
-    public void mergeAndNotifyStatusHasChanged(ChargeEntity chargeEntity) {
+    public void notifyStatusHasChanged(ChargeEntity chargeEntity) {
```
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/dao/ChargeDao.java]: Why this change? is it not merging too?
reading notifyStatusHasChange means to me only notify (and not update) 
(Maybe the super.merge is not necessary since the transaction boundary changed?)

File: src/main/java/uk/gov/pay/connector/model/AuthorisationRequest.java [hunk 2]
```diff
@@ -4,7 +4,7 @@
 import uk.gov.pay.connector.model.domain.ChargeEntity;
 import uk.gov.pay.connector.model.domain.GatewayAccountEntity;
 
-public class AuthorisationRequest {
+public class AuthorisationRequest implements GatewayRequest {
```
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/model/AuthorisationRequest.java]: :+1:

File: src/main/java/uk/gov/pay/connector/service/CardCaptureService.java [hunk 3]
```diff
@@ -2,106 +2,65 @@
...
-                    if (postCapture.isLeft())
-                        return left(postCapture.left().value());
-
-                    return right(captured.right().value());
-                };
-
         return chargeDao
                 .findByExternalId(chargeId)
-                .map(doCapture)
+                .map(TransactionalGatewayOperation.super::executeGatewayOperationFor)
```
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/service/CardCaptureService.java]: :+1:

File: src/main/java/uk/gov/pay/connector/service/TransactionalGatewayOperation.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import uk.gov.pay.connector.model.domain.ChargeEntity;
+
+import javax.persistence.OptimisticLockException;
+
+import static fj.data.Either.left;
+import static fj.data.Either.right;
+import static java.lang.String.format;
+import static uk.gov.pay.connector.model.ErrorResponse.conflictError;
+
+interface TransactionalGatewayOperation {
```
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/service/TransactionalGatewayOperation.java]: - This is clearly a Template Method pattern and it seems the intention is not for executeGatewayOperationFor to be override. It looks to me that _TransactionalGatewayOperation_ should be an abstract class and _executeGatewayOperationFor_ a public final method (a template method).

File: src/test/java/uk/gov/pay/connector/unit/service/CardCancelServiceTest.java [hunk 5]
```diff
@@ -31,24 +31,26 @@ public void shouldCancelACharge() throws Exception {
         Long accountId = 1L;
 
         ChargeEntity charge = createNewChargeWith(chargeId, ENTERING_CARD_DETAILS);
+        ChargeEntity reloadedCharge = mock(ChargeEntity.class);
 
-        when(mockedChargeDao.findByExternalIdAndGatewayAccount(charge.getExternalId(), accountId)).thenReturn(Optional.of(charge));
-        when(mockedAccountDao.findById(charge.getGatewayAccount().getId())).thenReturn(Optional.of(charge.getGatewayAccount()));
+        when(mockedChargeDao.findByExternalIdAndGatewayAccount(charge.getExternalId(), accountId))
+                .thenReturn(Optional.of(charge));
+        when(mockedChargeDao.merge(any()))
```
Comment (rauligs) [inline src/test/java/uk/gov/pay/connector/unit/service/CardCancelServiceTest.java]: I do prefer not the use of _any()_ but to know which element is being passed (could be done in a verify too). This applies anywhere where _any()_ is being used.

File: src/main/java/uk/gov/pay/connector/service/TransactionalGatewayOperation.java (hunk 4 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/TransactionalGatewayOperation.java]: That was also what I was leaning towards, in fact this is how it was implemented first - Imran has a different view point regarding this and has a few arguments against it - maybe have a quick chat with him about it. Personally I am happy either way.

Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/service/TransactionalGatewayOperation.java]: Anyway is much much cleaner and nice to see! :+1: 
I'm happy to merge with this approach and have a discussion if necessary afterwards.

File: src/main/java/uk/gov/pay/connector/dao/ChargeDao.java (hunk 1 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/dao/ChargeDao.java]: Indeed the merge should not be needed although looking into this it appears that there is still one place where it is necessary - this could be refactored however we decided not to spend time in this since the Auditing is going to be improved in the near future making this whole method redundant.

Renaming back to its original name for clarity.

File: src/test/java/uk/gov/pay/connector/unit/service/CardCancelServiceTest.java (hunk 5 above)
Comment (ckalista) [inline src/test/java/uk/gov/pay/connector/unit/service/CardCancelServiceTest.java]: Updated accordingly.

**Comment by ckalista** [comment]
rebuild
